File: src/js/environment.js
const DEAD = 0;
const ALIVE = 1;
const MIN_COUNT = 2;
const MAX_COUNT = 3;

export function getNextState(currentState, width, height) {
  const bounds = createBounds(width, height);
  let index = 0,
    nextState = new Uint8Array(width*height),
    maxHeight = height - 1;

  for (let line = 0; line < height; line++) {
    for (let column = 0; column < width; column++) {
      nextState[index] = getCellStatus(
        currentState[index],
        getNeighboursCount(currentState, column, line, maxHeight, bounds)
      );

      index++;
    }
  }

  return nextState;
}

export function getNeighboursCount(currentState, column, line, height, bounds) {
  let previousLine = line === 0 ? height : line - 1,
      nextLine = line === height ? 0 : line + 1;

  return getLineCount(currentState, column, bounds[line])
    + getLineCount(currentState, column, bounds[nextLine])
    + getLineCount(currentState, column, bounds[previousLine])
    - currentState[bounds[line].left + column];
}

function getLineCount(currentState, column, bounds) {
  let offset = bounds.left + column;

  return currentState[offset]
    + currentState[offset === bounds.left ? bounds.right : offset - 1]
    + currentState[offset === bounds.right ? bounds.left : offset + 1];
}

export function createBounds(width, height) {
  let bounds = [], leftBound;

  for (let i = 0; i < height; i++) {
    leftBound = i * width;
    bounds.push({
      left: leftBound,
      right: leftBound + width - 1
    });
  }

  return bounds;
}

export function getCellStatus (status, neighboursCount) {
  switch (neighboursCount) {
    case MAX_COUNT:
      return ALIVE;
    case 2:
      return status;
    default:
      return DEAD;
  }
}

File: src/js/c/environment.c
#include <stdlib.h>
#include <stdio.h>
#include <emscripten.h>
#include <string.h>

int dead = 0;
int alive = 1;

int minCount = 2;
int maxCount = 3;

int getCellStatus(int status, int neighboursCount) {
	switch(neighboursCount) {
		case 3: 
			status = alive;
			break;
		case 2:
			break;
		default:
			status = dead;
	}

	return status;
}


int getLineCount(char *current, int column, int boundLeft, int boundRight) {
	int offset = boundLeft + column;
	int left = (offset == boundLeft) ? current[boundRight] : current[offset - 1];
	int right = (offset == boundRight) ? current[boundLeft] : current[offset + 1]; 

	return current[offset] + left + right;
}

int getNeighboursCount(char *current, int column, int line, int height, int width, char *bounds) {
	int previousLine = (line == 0) ? height : line - 1;
	int nextLine = (line == height) ? 0 : line + 1;

  	return getLineCount(current, column, bounds[line], bounds[line + width])
      	+ getLineCount(current, column, bounds[nextLine], bounds[nextLine + width])
      	+ getLineCount(current, column, bounds[previousLine], bounds[previousLine + width])
      	- current[bounds[line] + column];
}

void createBounds(char *bounds, int width, int height) {
	int leftBound = 0;

	for (int i = 0; i < height; i++) {
		leftBound = i * width;
		bounds[i] = leftBound;
		bounds[i + width] = leftBound + width - 1;
	}

	return;
}

EMSCRIPTEN_KEEPALIVE
void getNextState(char *current, char *next, char *bounds, int width, int height) {
	createBounds(bounds, width, height);
	int maxHeight = height - 1;
	int index = 0;

	for (int i = 0; i < height; i++) {
		for (int j = 0; j < width; j++) {
			next[index] = getCellStatus(
				current[index],
				getNeighboursCount(current, j, i, maxHeight, width, bounds)
			);

			index++;
		}
	}

	return;
}